ttl.py

```diff
@@ -0,0 +1,153 @@
+        channel: Monitoring TTL channel number.
+        monitor_type: Monitoring value type.
+    """
+    channel: int
+    monitor_type: MonitorType
+
+    def __hash__(self) -> int:
+        """Overridden."""
+        return hash(f"{str(self.channel)}_{self.monitor_type.value}")
+
+
+# {"MonitorType value": {"TTL device name": "modified value"}}
+Modifications = dict[str, dict[str, bool]]
+ModificationQueue = SortedQueue[float, StatusType]
+
+
+class TTLManager:
+    """Manages the connection to ARTIQ moninj proxy and TTL status.
+
+    Attributes:
+        connection: CommMonInj instance for connection to ARTIQ moninj proxy.
+        queue: SortedQueue with modified StatusType.
+        values: Dictionary whose keys are StatusType and values are modified values.
+        modified: Event set when any value is modified.
+    """
+
+    def __init__(self):
+        self.connection = CommMonInj(self.monitor_cb, self.injection_status_cb)
+        self.queue = ModificationQueue()
+        self.values: dict[StatusType, bool] = {}
+        self.modified = asyncio.Event()
+
+    async def connect(self, core_addr: str, ttl_devices: list[str]):
+        """Connects to ARTIQ moninj proxy.
+        
+        Args:
+            See main.configs.
+        """
+        await self.connection.connect(core_addr)
+        for device in ttl_devices:
+            channel = device_to_channel[device]
+            self.connection.monitor_probe(1, channel, TTLProbe.level.value)
+            self.connection.monitor_injection(1, channel, TTLOverride.level.value)
+            self.connection.monitor_injection(1, channel, TTLOverride.en.value)
+
+    def current_status(self, devices: list[str]) -> tuple[float, Modifications]:
+        """Returns the current timestamp and status.
+        
+        Args:
+            devices: List of target TTL device names.
+        """
+        modifications = {ty.value: {} for ty in MonitorType}
+        for device in devices:
+            channel = device_to_channel[device]
+            for ty in MonitorType:
+                modifications[ty.value][device] = self.values[StatusType(channel, ty)]
+        return time.time(), modifications
+
+    def modifications_since(
+        self, devices: list[str], timestamp: float
+    ) -> tuple[float, Modifications]:
+        """Returns the latest timestamp and modifications since the given timestamp.
+        
+        Args:
+            devices: List of target TTL device names.
+            timestamp: Timestamp of the latest update.
+        """
+        modifications = {ty.value: {} for ty in MonitorType}
+        latest, modification_types = self.queue.tail(timestamp)
+        filtered_modification_types = filter(
+            lambda ty: channel_to_device[ty.channel] in devices, set(modification_types)
+        )
+        for ty in filtered_modification_types:
+            device = channel_to_device[ty.channel]
+            modifications[ty.monitor_type.value][device] = self.values[ty]
```

Comment:
In the for loop, it needs the `device` anyway. How about not filtering and checking it in the for loop, using `continue`?

Reply:
It can reduce the duplicated for loop! I will apply it.